subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/Test.java

```diff
@@ -820,10 +824,15 @@ public Test setExcludes(Iterable<String> excludes) {
 
     @Internal
     public TestFramework getTestFramework() {
```

Comment:
I talked to Adam about my misgivings on this and I think there are a couple of things...

- The default/convention for determining the test framework to use is in the Test task itself vs pushing out to the plugins.
- The way this is wired into up-to-date checks looks brittle.  This method just happens to get called in the right place (`getOptions()` below) to cause the testFramework to be set. 
- The way we detect the test framework is brittle as well (based on file names).  
- I think this also pushes us further away from something higher-level from deciding which test framework to use.  e.g., you could imagine a plugin that would add the JUnit dependency and could set the test framework too.

I've pushed this commit to a separate branch: 
https://github.com/gradle/gradle/commit/86a1e971699d20d4855c626e9746cf508fc2716e
- On Adam's suggestion, I changed testFramework into a `Property<TestFramework>` and exposed it as another getter. I didn't do it, but this getter could be marked as `@Nested` and the TestFramework subclasses could be annotated properly to make it all work. This would make it clearer that the TestFramework was an input of some kind and prevent anything from sneaking in as an input. I don't think we need to do this now.
- I pulled out the details for creating TestFramework instances into a factory and moved the way the conventional TestFramework is decided.  In each of our plugins that create Test tasks, I applied some convention.  In JavaBasePlugin, the convention is to use JUnit (what we do now).  In JavaPlugin, the convention is to use the test framework detection to pick a test framework.  In the JUnit software model plugin, the convention is to use JUnit always.
- To avoid sniffing file names, I also changed the test framework detection to look at the _direct_ dependencies of the test's implementation.  I don't think this is as good as it could be (since I see multiple JUnit platform dependencies that look roughy equivalent), but it follows a simple rule: if JUnit platform is available, use JUnit platform.  If TestNG is available, use TestNG. if JUnit is available, use JUnit.  If nothing is available, assume JUnit (for backwards compatibility). 

I think this alleviates my initial concerns:
- We can't resolve dependencies accidentally during configuration
- We avoid inspecting file names 
- We don't make the Test task more complicated (it just does whatever testFramework says).

My commit breaks some tests, but it seems to work mostly.